fe/fe-core/src/main/java/com/starrocks/scheduler/mv/MVPCTRefreshListPartitioner.java

```diff
@@ -406,6 +406,89 @@ public void filterPartitionByRefreshNumber(Set<String> mvPartitionsToRefresh,
+        }
+
+        // filter by partition ttl
+        filterPartitionsByTTL(partitionToCells, false);
+        if (CollectionUtils.sizeIsEmpty(partitionToCells)) {
+            return;
+        }
+        Map<String, PListCell> toRefreshPartitions = partitionToCells.entrySet().stream()
+                .collect(Collectors.toMap(Map.Entry::getKey, e -> (PListCell) e.getValue()));
+
+        Map<String, Map<Table, Set<String>>> mvToBaseNameRefs = mvContext.getMvRefBaseTableIntersectedPartitions();
+
+        // TODO: Sort by List Partition's value is weird because there maybe meaningless or un-sortable,
+        // users should take care of `partition_ttl_number` for list partition.
+        LinkedHashMap<String, PListCell> sortedPartition = toRefreshPartitions.entrySet().stream()
+                .sorted((e1, e2) -> e2.getValue().compareTo(e1.getValue())) // reverse order(max heap)
+                .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue, (e1, e2) -> e1, LinkedHashMap::new));
+        Iterator<String> toSelectedPartitionNameIter = sortedPartition.keySet().iterator();
+
+        // dynamically obtain the number of partitions to be refreshed this time
+        int adaptivePartitionRefreshNumber = getAdaptivePartitionRefreshNumber(toSelectedPartitionNameIter);
+        Iterator<Map.Entry<String, PListCell>> iter = sortedPartition.entrySet().iterator();
+        // iterate adaptivePartitionRefreshNumber times
+        for (int i = 0; i < adaptivePartitionRefreshNumber; i++) {
+            if (iter.hasNext()) {
+                iter.next();
+            }
+        }
+
+        // compute next partition list cells in the next task run
+        Set<PListCell> nextPartitionValues = Sets.newHashSet();
+        while (iter.hasNext()) {
+            Map.Entry<String, PListCell> entry = iter.next();
+            nextPartitionValues.add(entry.getValue());
+            // remove the partition which is not reserved
+            toRefreshPartitions.remove(entry.getKey());
+        }
+        logger.info("Filter partitions by adaptive_partition_number, ttl_number:{}, result:{}, remains:{}",
+                adaptivePartitionRefreshNumber, toRefreshPartitions, nextPartitionValues);
+        // do filter input mvPartitionsToRefresh since it's a reference
+        mvPartitionsToRefresh.retainAll(toRefreshPartitions.keySet());
+        if (CollectionUtils.isEmpty(nextPartitionValues)) {
+            return;
+        }
+        if (!tentative) {
+            // partitionNameIter has just been traversed, and endPartitionName is not updated
+            // will cause endPartitionName == null
+            mvContext.setNextPartitionValues(PListCell.batchSerialize(nextPartitionValues));
+        }
+    }
+
+    private int getAdaptivePartitionRefreshNumber(Iterator<String> partitionNameIter) {
+
+        Map<String, Map<Table, Set<String>>> mvToBaseNameRefs = mvContext.getMvRefBaseTableIntersectedPartitions();
```

Comment:
switch to `default` policy if base tables contain non-olap-table  to avoid `Partition partition = table.getPartition(partitionName);` exception?

And fix this in the later pr?

Reply:
I have added a check before selecting the refresh mode. If the base tables of the materialized view include any external tables, only the default refresh mode will be used. I will include this fix in the next PR.